addons/hr_holidays/models/resource.py

```diff
@@ -30,6 +33,123 @@ def _check_compare_dates(self):
+            leave.write({
+                'date_from': leave_tz.localize(gto_end + timedelta(seconds=1))\
+                        .astimezone(UTC).replace(tzinfo=None)
+            })
+            return self.env['hr.leave'].sudo()
+        if gto_start > leave_start\
+                and gto_end < leave_end:
+            copys = {
+                'date_from': leave.date_from,
+                'date_to': leave_tz.localize(gto_start - timedelta(seconds=1))\
+                        .astimezone(UTC).replace(tzinfo=None)
+            }
+            leave.write({
+                'date_from': leave_tz.localize(gto_end + timedelta(seconds=1))\
+                        .astimezone(UTC).replace(tzinfo=None)
+            })
+            return leave.copy(copys)
+        if gto_start > leave_start\
+                and gto_start < leave_end\
+                and gto_end >= leave_end:
+            leave.write({
+                'date_to': leave_tz.localize(gto_start - timedelta(seconds=1))\
+                        .astimezone(UTC).replace(tzinfo=None)
+            })
+            return self.env['hr.leave'].sudo()
+
+    def _split_leave(self, leave, time_domain_dict):
+        new_leaves = self.env['hr.leave'].sudo()
+        for record in sorted(
+                filter(lambda r: r['date_to'] > leave.date_from and r['date_from'] < leave.date_to, time_domain_dict),
+                key=lambda r: r['date_from']):
+            new_leave = self._split_leave_on_gto(leave, record)
+            if new_leave:
+                new_leaves |= new_leave
+        return new_leaves
+
+    def _reevaluate_leaves(self, time_domain_dict):
+        if not time_domain_dict:
+            return
+
+        domain = self._get_domain(time_domain_dict)
+        leaves = self.env['hr.leave'].search(domain)
+        if not leaves:
+            return
+
+        previous_durations = leaves.mapped('number_of_days')
+        previous_states = leaves.mapped('state')
+        leaves.sudo().write({
+            'state': 'draft',
+        })
+        self.env.add_to_compute(self.env['hr.leave']._fields['number_of_days'], leaves)
+        sick_time_status = self.env.ref('hr_holidays.holiday_status_sl')
+        for previous_duration, leave, state in zip(previous_durations, leaves, previous_states):
+            duration_difference = previous_duration - leave.number_of_days
+            if duration_difference > 0 and leave['holiday_allocation_id'] and leave.number_of_days == 0.0:
+                message = _("Due to a change in global time offs, you have been granted %s day(s) back.", duration_difference)
+                leave._notify_change(message)
+            if leave.number_of_days > previous_duration\
+                    and leave.holiday_status_id not in sick_time_status:
+                new_leaves = self._split_leave(leave, time_domain_dict)
+                leaves |= new_leaves
+                previous_states += [state for i in range(len(new_leaves))]
```

Comment:
```suggestion
                previous_states += [state] * len(new_leaves)
```
Or keep it like that, as you prefer :innocent: